Flaky ReaderPoolTest.testGetAndCloseRace test

### Describe the bug

Flapping test:

```
2021-12-28T09:44:59.4359640Z 2021-12-28T09:44:58.884459Z I i.q.c.AbstractCairoTest Tearing down test ReaderPoolTest#testGetAndCloseRace
2021-12-28T09:44:59.4362640Z 2021-12-28T09:44:58.902433Z E i.q.TestListener ***** Test Failed *****io.questdb.cairo.pool.ReaderPoolTest.testGetAndCloseRace : 
2021-12-28T09:44:59.4364750Z java.lang.AssertionError: [15,12,16,17,18] expected:<3> but was:<6>
2021-12-28T09:44:59.4366400Z    at org.junit.Assert.fail(Assert.java:89)
2021-12-28T09:44:59.4367700Z    at org.junit.Assert.failNotEquals(Assert.java:835)
2021-12-28T09:44:59.4369500Z    at org.junit.Assert.assertEquals(Assert.java:647)
2021-12-28T09:44:59.4374600Z    at io.questdb.test.tools.TestUtils.assertMemoryLeak(TestUtils.java:418)
2021-12-28T09:44:59.4376290Z    at io.questdb.cairo.pool.ReaderPoolTest.assertWithPool(ReaderPoolTest.java:877)
2021-12-28T09:44:59.4378090Z    at io.questdb.cairo.pool.ReaderPoolTest.assertWithPool(ReaderPoolTest.java:873)
2021-12-28T09:44:59.4380670Z    at io.questdb.cairo.pool.ReaderPoolTest.testGetAndCloseRace(ReaderPoolTest.java:382)
2021-12-28T09:44:59.4384810Z    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2021-12-28T09:44:59.4386240Z    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2021-12-28T09:44:59.4387810Z    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2021-12-28T09:44:59.4393480Z    at java.lang.reflect.Method.invoke(Method.java:566)
2021-12-28T09:44:59.4395510Z    at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
```


Observed in https://dev.azure.com/questdb/questdb/_build/results?buildId=4386&view=logs&j=3fa4b4be-6db3-54fa-4432-6a83730bebb7&t=bb90c814-cfaa-5557-8498-4851227a92fe

Please do not confuse with https://github.com/questdb/questdb/issues/1513, which is fixed

### To reproduce

Run this test to failure

### Expected Behavior

_No response_

### Environment

```markdown
- **QuestDB version**:
- **OS**:
- **Browser**:
```


### Additional context

_No response_

--- a/core/src/main/java/io/questdb/cairo/pool/ReaderPool.java
+++ b/core/src/main/java/io/questdb/cairo/pool/ReaderPool.java
@@ -242,7 +242,6 @@ public class ReaderPool extends AbstractPool implements ResourcePool<TableReader
                             Unsafe.arrayPutOrdered(e.allocations, i, UNALLOCATED);
                         } else {
                             casFailures++;
-
                             if (deadline == Long.MAX_VALUE) {
                                 r.goodby();
                                 LOG.info().$("shutting down. '").$(r.getTableName()).$("' is left behind").$();
@@ -256,7 +255,7 @@ public class ReaderPool extends AbstractPool implements ResourcePool<TableReader
         }
 
         // when we are timing out entries the result is "true" if there was any work done
-        // when we closing pool, the result is true when pool is empty
+        // when we're closing pool, the result is true when pool is empty
         if (closeReason == PoolConstants.CR_IDLE) {
             return removed;
         } else {
@@ -317,12 +316,7 @@ public class ReaderPool extends AbstractPool implements ResourcePool<TableReader
 
             e.releaseTimes[index] = clock.getTicks();
             Unsafe.arrayPutOrdered(e.allocations, index, UNALLOCATED);
-
-            // todo: there is a race condition between this method and
-            //   releaseAll() when the latter shuts down the pool. I thought of adding a version counter
-            //   that each method will attempt to increment thus recognising race condition and
-            //   taking action
-            return true;
+            return !isClosed();
         }
 
         LOG.error().$('\'').$(name).$("' is available [at=").$(e.index).$(':').$(index).$(']').$();

--- a/core/src/main/java/io/questdb/cairo/pool/ReaderPool.java
+++ b/core/src/main/java/io/questdb/cairo/pool/ReaderPool.java
@@ -97,7 +97,7 @@ public class ReaderPool extends AbstractPool implements ResourcePool<TableReader
 
                     if (isClosed()) {
                         e.readers[i] = null;
-                        r.goodby();
+                        r.goodbye();
                         LOG.info().$('\'').utf8(name).$("' born free").$();
                         return r;
                     }
@@ -243,7 +243,7 @@ public class ReaderPool extends AbstractPool implements ResourcePool<TableReader
                         } else {
                             casFailures++;
                             if (deadline == Long.MAX_VALUE) {
-                                r.goodby();
+                                r.goodbye();
                                 LOG.info().$("shutting down. '").$(r.getTableName()).$("' is left behind").$();
                             }
                         }
@@ -266,7 +266,7 @@ public class ReaderPool extends AbstractPool implements ResourcePool<TableReader
     private void closeReader(long thread, Entry entry, int index, short ev, int reason) {
         R r = entry.readers[index];
         if (r != null) {
-            r.goodby();
+            r.goodbye();
             r.close();
             LOG.info().$("closed '").utf8(r.getTableName())
                     .$("' [at=").$(entry.index).$(':').$(index)
@@ -360,11 +360,14 @@ public class ReaderPool extends AbstractPool implements ResourcePool<TableReader
                 if (pool != null && entry != null && pool.returnToPool(this)) {
                     return;
                 }
-                super.close();
+                final Entry e = this.entry;
+                if (e == null || Unsafe.cas(e.allocations, index, UNALLOCATED, Thread.currentThread().getId())) {
+                    super.close();
+                }
             }
         }
 
-        private void goodby() {
+        private void goodbye() {
             entry = null;
             pool = null;
         }

